### DISCUSSION TITLE
Using Replay to Solve Hydration Errors

### DISCUSSION THREAD

**Discussion by scefali** [getsentry/sentry#62097] [Product Spitballin']
## Introduction

[Hydration Errors](https://nextjs.org/docs/messages/react-hydration-error) are quite common yet hard to fix. They happen when the HTML in the first render cycle doesn't match the HTML sent by the server. It causes both a full re-render can sometimes can cause adverse side effects. Currently, hydration errors are being filtered out by Sentry's inbound filters because Sentry cannot give helpful debugging information and they are very noisy. **Our goal is to make hydration errors actionable so developers can actually fix them.**

## Using Session Replay

![image](https://github.com/getsentry/sentry/assets/1400464/b92d78cd-2d37-44af-8cd2-f93e79eb0d99)

We can use session replays to display a visual and html diff of when the app starts up. Because we're filtering out the error events we've added an additional `replay.hydrate-error` breadcrumb to the Replay SDK to record when these errors occur, similar to a click event.

Once we have the hydration error breadcrumb, we can get the initial pageload and compare it with the html state at the time of the `replay.hydrate-error` event. Comparing these visually and using an html diff it should point people into the right direction of the hydration error's source.

**We'd love to hear from you on:**

1. How well a solution like this would help you debug hydration errors?
2. Whether you have any other suggestions for how we can improve the experience around fixing hydration errors?

## Getting Started
See this changelog for notes on how to get started https://changelog.getsentry.com/announcements/diff-hydration-errors-with-replay

**Comment by brettdh** [comment]
This would be _so lovely_ to have... in Remix. 😅

**Reply by scttcper** [reply]
@brettdh there's a good chance it already works with remix if you have a fully updated sdk. [See this changelog](https://changelog.getsentry.com/announcements/diff-hydration-errors-with-replay) for how to get started

**Reply by brettdh** [reply]
Finally got a chance to test this, as I have a real hydration error that I'm struggling with. After upgrading `@sentry/remix` to latest (`7.113.0`) and reproducing the error, here's what I get:

<img width="1594" alt="Sentry wasn't able to identify the correct event to display a diff for this hydration error." src="https://github.com/getsentry/sentry/assets/91550/97cba5dd-1b79-4527-859b-6e9b59b26f8f">

Both the HTML and visual diffs look identical. Should I open a bug, or is there a setup step I'm missing?

**Reply by scttcper** [reply]
there's a few possibilities:
- the replay didn't start early enough to capture the initial html (more common in localhost). Race between when the replay starts recording and when the hydration error happens.
- the hydration error is in the header which isn't currently shown in the diff
- if you have maskAllText enabled, maybe the hydration error was in the text?

you can also open an issue or link here. if you're using cloudflare there's a good chance one of the settings is changing your headers

**Reply by brettdh** [reply]
Thanks for the quick response. To those possibilities:
- I don't currently have reporting to Sentry enabled on localhost; these errors are reported from Remix running in AWS Lambda + API Gateway. If it's a race condition, though, could I temporarily insert a sleep somewhere to avoid the race? If so, where?
- Hmm, I don't think I was aware that a difference in headers could cause a hydration error, but I also don't think my current changes could have caused a change in the headers. Also, I do see the same hydration error on localhost (though I don't report it to Sentry in that case), which I think eliminates the possibility of AWS header-munging being the cause of the hydration error.
- I have `maskAllText` set to `false`.

**Comment by scttcper** [comment]
We've published a changelog with how to get started debugging hydration errors using session replay https://changelog.getsentry.com/announcements/diff-hydration-errors-with-replay

**Comment by brysongilbert** [comment]
This has the makings of a super useful feature, thank you! 🙏 One thing we have noticed at a glance is that differences in attribute order are highlighted in the diff. For example:

![Screenshot 2024-06-26 at 14 02 14](https://github.com/getsentry/sentry/assets/997133/ddd22a94-475a-4d47-9dc0-7feb4c1295c4)

(Note that this is just a one-line snippet of a much larger diff. And in this case the particular line comes from framework code, not app code.)

I stand to be corrected on this but I don't believe React cares about attribute order, in terms of a hydration mismatch. So you get a little visual noise from these if you're looking at a relatively larger/longer diff.

Wish list item: for only missing/new attributes or differing attribute values, to be highlighted, and for attribute order (where the attribute and value are otherwise identical) to be ignored.

**Reply by norbert-777** [reply]
Hi, This issue still exists and it creates unnecessary noise. Is there any chance that Hydration error would be improved?

**Comment by felixmosh** [comment]
For some reason, the "Before hydaration" is always empty...
I'm using Next.js v13
![image](https://github.com/user-attachments/assets/e8416854-441b-4ed3-8270-3e133d59344d)
Using the cdn bundle script in the head of the page.
```html
<script src="https://browser.sentry-cdn.com/8.20.0/bundle.replay.min.js" integrity="sha384-3RyGu38xD2r+OrNF2g9Ekem9LqdE766XjQCz0pIzabGm4T52CIDxkyWUq7klKDVd" crossorigin="anonymous" defer=""></script>
```

Any idea why?

**Reply by scttcper** [reply]
@felixmosh I haven't investigated using the cdn bundle for this, i imagine the defer property at the very least would impact the ability to load soon enough to capture the "before"

**Reply by felixmosh** [reply]
This is weird...since the the lib does capture the hydration phase... No?

**Comment by ryan953** [comment]
Good news everyone, we've moved this feature into GA today! Every Sentry org will be able to install the Replay integration and have Hydration Errors captured automatically. Make sure you're on SDK version 7.87.0 or higher.

https://sentry.io/changelog/debug-hydration-errors-with-our-diff-tool---now-generally-available/

**Comment by imteekay** [comment]
I'm instrumenting Sentry Replay on my web app and was trying to configure for our specific case.
Is it possible to:
1. Configure Replay to capture only hydration errors and skip other errors?
2. Send hydration errors to Sentry (so I can see on Replay) without impacting the availability metric (Crash Free Sessions)?

For 1, I thought of using `beforeErrorSampling` from Replay config but I'm not sure if there's a better way of doing this. Something like this:

```ts
// https://reactjs.org/docs/error-decoder.html/?invariant=418
const ERROR_418 =
  /Hydration failed because the initial UI does not match what was rendered on the server/;
const ERROR_418_MINIFIED = /Minified React error #418/;

function isReactHydrationError(error: string): boolean {
  return (
    ERROR_418.test(error) ||
    ERROR_418_MINIFIED.test(error)
    // ... and other hydration errors
  );
}

function getIntegrations() {
  if (typeof window === 'undefined') {
    return [];
  }

  return [
    replayIntegration({
      maskAllText: true,
      blockAllMedia: true,
      beforeErrorSampling: (event) => {
        const error = event.exception?.values?.[0]?.value;
        return isReactHydrationError(error);
      },
    }),
  ];
}
```

For 2, I thought of updating the `mechanism` value from the exception error to `handled = true`. Does that work? It would look like this:

```ts
const error = event.exception?.values?.[0]?.value;

if (isReactHydrationError(error)) {
  event.exception.values.forEach((e) => {
    e.mechanism.handled = false;
  });
}
```
